src/timely-util/src/builder_async.rs

```diff
@@ -62,6 +57,55 @@ pub struct OperatorBuilder<G: Scope> {
     shutdown_button: Button,
 }
 
+/// A helper trait abstracting over an input handle. It facilitates keeping around type erased
+/// handles for each of the operator inputs.
+trait InputQueue<T: Timestamp> {
+    /// Accepts all available input into local queues.
+    fn accept_input(&mut self);
+
+    /// Drains all available input and empties the local queue.
+    fn drain_input(&mut self);
+
+    /// Registers a frontier notification to be delivered.
+    fn notify_progress(&mut self, upper: Antichain<T>);
+}
+
+impl<T, D, C, P> InputQueue<T> for InputHandleQueue<T, D, C, P>
+where
+    T: Timestamp,
+    D: Container,
+    C: InputConnection<T> + 'static,
+    P: Pull<BundleCore<T, D>> + 'static,
+{
+    fn accept_input(&mut self) {
+        let mut queue = self.queue.borrow_mut();
+        while let Some((cap, data)) = self.handle.next() {
+            let cap = self.connection.accept(cap);
+            queue.push_back(Event::Data(cap, data.take()));
```

Comment:
I think for places that were previously using `next_mut` we have the same buffer reuse because this `.take()` here swaps in a buffer. There will be a small regression for operators that were using `.next()` which served shared references and in this case `data.take()` will clone the data, but we don't have any operators that used `.next()`. So overall I think we're in a good place. We could revisit this if it ends up being problematic

Reply:
same!